docs/references/security/rust-canister-development-security-best-practices.md

```diff
@@ -183,43 +183,179 @@ A canister could be rendered unusable so it could never be upgraded again e.g. d
+
+We fist provide a few definitions. A _call_ is a canister's implementation of either an [update](https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-call) or [query call](https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-query) that it exposes. For example, if the Rust CDK is used, these are usually annotated with `#[query]` or `#[update]`, respectively. A _message_ is a set of consecutive instructions that a subnet executes for a canister. We'll see in the following that a call can be split into several messages if inter-canister calls are made. The following properties are essential: 
+
+**Property 1**: Only a single message is processed at a time in a subnet. So message execution is sequential, and never parallel.
+
+**Property 2**: Each call (query / update) triggers a message. When an inter-canister call is made, the code after the call (the callback, highlighted in blue) is executed as a separate message. 
+
+For example, consider the following Motoko code: 
+
+
+![example_highlighted_code](_attachments/example_highlighted_code.png)
+
+The first message that is executed here are lines 2-3, until the inter-canister call is made using the `await` syntax (orange box). The second message executes lines 3-5: when the inter-canister call returns (blue box). We call this part the _callback_ of the inter-canister call. The two messages involved in this example will always be scheduled sequentially. 
+
+**Property 3**: Messages from interleaving calls have no reliable execution ordering.
+
+Let's consider the above example code again, and assume the method `example` is called twice in parallel, the resulting calls being Call 1 and Call 2. The following illustration shows two possible message orderings. On the left, the first call's messages are scheduled first, and only then the second call's messages are executed. On the right, we see another possible message scheduling, where the first messages of each call are executed first. Your code should result in a correct state regardless of the message ordering.
+
+![example_orderings](_attachments/example_orderings.png)
+
+**Property 4**: On a trap / panic, modifications to the canister state for the current message are not applied.
+
+For example, if a trap in the second message (blue box) of the above example occurs, canister state changes resulting from that message, even earlier in the blue box, are discarded. However, note that any state changes from earlier messages and in particular the first message (orange box) have been applied, as that message executed successfully. 
+
+For more details, refer to the Interface Specification [section on ordering guarantees](https://internetcomputer.org/docs/current/references/ic-interface-spec/#ordering_guarantees) and the section on [abstract behaviour](https://internetcomputer.org/docs/current/references/ic-interface-spec/#abstract-behavior) which defines message execution in more detail. 
+
```

Comment:
```suggestion

**Property 5**: Inter-canister messaging is not reliable.

Every inter-canister call is guaranteed to receive a response. However, the response does not have to be successful, but can also be a reject response. The reject may come from the called canister, but it may also be generated by the Internet Computer. Such system-generated rejects can occur at any time, and cannot be controlled by canister authors. Thus, it's important that the calling canister handles reject responses as well. Note that a reject response generally guarantees that the message had not been processed by the called canister.
```

Reply:
Nice, added it as suggested. Thanks!